core/shared/src/main/scala/scalaz/zio/ZIO.scala

```diff
@@ -1632,226 +1662,154 @@ private[zio] trait ZIOFunctions extends Serializable {
    */
-  final def reduceAllPar[R >: LowerR, R1 >: LowerR <: R, E <: UpperE, A](a: ZIO[R, E, A], as: Iterable[ZIO[R1, E, A]])(
+  final def reduceAllPar[R, R1 <: R, E, A](a: ZIO[R, E, A], as: Iterable[ZIO[R1, E, A]])(
     f: (A, A) => A
   ): ZIO[R1, E, A] =
     as.foldLeft[ZIO[R1, E, A]](a) { (l, r) =>
       l.zipPar(r).map(f.tupled)
     }
 
   /**
-   * Merges an `Iterable[IO]` to a single IO, working sequentially.
-   */
-  final def mergeAll[R >: LowerR, E <: UpperE, A, B](
-    in: Iterable[ZIO[R, E, A]]
-  )(zero: B)(f: (B, A) => B): ZIO[R, E, B] =
-    in.foldLeft[ZIO[R, E, B]](succeedLazy[B](zero))((acc, a) => acc.zip(a).map(f.tupled))
-
-  /**
-   * Merges an `Iterable[IO]` to a single IO, working in parallel.
+   * Requires that the given `IO[E, Option[A]]` contain a value. If there is no
+   * value, then the specified error will be raised.
    */
-  final def mergeAllPar[R >: LowerR, E <: UpperE, A, B](
-    in: Iterable[ZIO[R, E, A]]
-  )(zero: B)(f: (B, A) => B): ZIO[R, E, B] =
-    in.foldLeft[ZIO[R, E, B]](succeedLazy[B](zero))((acc, a) => acc.zipPar(a).map(f.tupled))
+  final def require[E, A](error: E): IO[E, Option[A]] => IO[E, A] =
+    (io: IO[E, Option[A]]) => io.flatMap(_.fold[IO[E, A]](fail[E](error))(succeed[A]))
 
   /**
-   * Strictly-evaluated unit lifted into the `ZIO` monad.
+   * Acquires a resource, uses the resource, and then releases the resource.
+   * However, unlike `bracket`, the separation of these phases allows
+   * the acquisition to be interruptible.
+   *
+   * Useful for concurrent data structures and other cases where the
+   * 'deallocator' can tell if the allocation succeeded or not just by
+   * inspecting internal / external state.
    */
-  final val unit: ZIO[Any, Nothing, Unit] = succeed(())
+  def reserve[R, E, A, B](reservation: ZIO[R, E, Reservation[R, E, A]])(use: A => ZIO[R, E, B]): ZIO[R, E, B] =
+    ZManaged(reservation).use(use)
 
   /**
-   * The moral equivalent of `if (p) exp`
+   * Returns an effect that accesses the runtime, which can be used to
+   * (unsafely) execute tasks. This is useful for integration with
+   * non-functional code that must call back into functional code.
    */
-  final def when[R >: LowerR, E <: UpperE](b: Boolean)(zio: ZIO[R, E, _]): ZIO[R, E, Unit] =
-    if (b) zio.unit else unit
+  final def runtime[R]: ZIO[R, Nothing, Runtime[R]] =
+    for {
+      environment <- environment[R]
+      platform    <- effectTotalWith(ZIO.identityFn[Platform])
+    } yield Runtime(environment, platform)
 
   /**
-   * The moral equivalent of `if (p) exp` when `p` has side-effects
+   * Sleeps for the specified duration. This method is asynchronous, and does
+   * not actually block the fiber.
    */
-  final def whenM[R >: LowerR, E <: UpperE](b: ZIO[R, E, Boolean])(zio: ZIO[R, E, _]): ZIO[R, E, Unit] =
-    b.flatMap(b => if (b) zio.unit else unit)
+  final def sleep(duration: Duration): ZIO[Clock, Nothing, Unit] =
```

Comment:
I would add the `sleep` method to `TaskR`, as well, because it is compatible with `TaskR` type.